clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -1183,20 +1183,42 @@ private TxnOffsetCommitHandler txnOffsetCommitHandler(TransactionalRequestResult
         return new TxnOffsetCommitHandler(result, builder);
     }
 
+    private void throwPendingTransitionError(String operation) {
+        throw new KafkaException("Cannot attempt operation `" + operation + "` "
```

Comment:
Intuitively, I would have raised an `IllegalStateException` here. Is there a reason not to?

Reply:
I also slightly favor `IllegalStateException`, but I used `KafkaException` for consistency with `transitionTo`.